src/Numerics/SystemSolvers/batched_generalized_minimal_residual_solver.jl

```diff
@@ -263,517 +276,381 @@ function BatchedGeneralizedMinimalResidual(
-@inline convert_structure!(x::MPIStateArray, y, reshape_tuple, permute_tuple) =
-    convert_structure!(x.data, y, reshape_tuple, permute_tuple)
-
-# Kernels
-"""
-    initialize_gmres_kernel!(gmres)
-
-# Description
-Initializes the gmres struct by calling
-- initialize_arnoldi
-- initialize_QR!
-- update_arnoldi!
-- update_QR!
-- solve_optimization!
-It is assumed that the first two krylov vectors are already constructed
-
-# Arguments
-- `gmres`: (struct) gmres struct
+@kernel function batched_initialize!(resnorms, g0, batched_krylov_basis)
+    @uniform begin
+        Ndof = size(batched_krylov_basis, 2)
+        Mp1 = size(g0, 2)
+        FT = eltype(batched_krylov_basis)
+    end
+    cidx = @index(Global)
 
-# Return
-(implicitly) kernel abstractions function closure
-"""
-@kernel function initialize_gmres_kernel!(gmres)
-    I = @index(Global)
-    initialize_arnoldi!(gmres, I)
-    update_arnoldi!(1, gmres, I)
-    initialize_QR!(gmres, I)
-    update_QR!(1, gmres, I)
-    solve_optimization!(1, gmres, I)
-end
+    # Initialize entire RHS storage
+    @inbounds for j in 1:Mp1
+        g0[cidx, j] = FT(0.0)
+    end
 
-"""
-    gmres_update_kernel!(i, gmres)
+    # Now we compute the first element of g0[cidx, :],
+    # which is determined by the column norm of the initial residual ∥r0∥_2:
+    # g0 = ∥r0∥_2 e1
+    @inbounds for j in 1:Ndof
+        g0[cidx, 1] +=
+            batched_krylov_basis[1, j, cidx] * batched_krylov_basis[1, j, cidx]
+    end
+    g0[cidx, 1] = sqrt(g0[cidx, 1])
 
-# Description
-kernel that calls
-- update_arnoldi!
-- update_QR!
-- solve_optimization!
-Which is the heart of the gmres algorithm
+    # Normalize the batched_krylov_basis by the (local) residual norm
+    @inbounds for j in 1:Ndof
+        batched_krylov_basis[1, j, cidx] /= g0[cidx, 1]
+    end
 
-# Arguments
-- `i`: (int) interation index
-- `gmres`: (struct) gmres struct
-- `I`: (int) thread index
+    # Record initialize residual norm in the column
+    resnorms[cidx] = g0[cidx, 1]
 
-# Return
-kernel object from KernelAbstractions
-"""
-@kernel function gmres_update_kernel!(i, gmres)
-    I = @index(Global)
-    update_arnoldi!(i, gmres, I)
-    update_QR!(i, gmres, I)
-    solve_optimization!(i, gmres, I)
+    nothing
 end
 
-"""
-    construct_solution_kernel!(i, gmres)
-
-# Description
-given step i of the gmres iteration, constructs the "best" solution of the linear system for the given Krylov subspace
-
-# Arguments
-- `i`: (int) gmres iteration
-- `gmres`: (struct) gmres struct
-
-# Return
-kernel object from KernelAbstractions
-"""
-@kernel function construct_solution_kernel!(i, gmres)
-    M, I = @index(Global, NTuple)
-    tmp = zero(eltype(gmres.b))
-    @inbounds for j in 1:i
-        tmp += gmres.Q[M, j, I] * gmres.sol[j, I]
+@kernel function batched_arnoldi_process!(
+    resnorms,
+    g0,
+    H,
+    Ω,
+    batched_krylov_basis,
+    j,
+)
+    @uniform begin
+        Ndof = size(batched_krylov_basis, 2)
```

Comment:
Is this something we want at compile time? Guessing it will be large, so probably not. (Just wondering.)

Reply:
I think I addressed this now. This information is freely available so I'm just adding it as an additional kernel argument.

Reply:
Thinking about this some more, I think that to remove the `convert_structure` function we will probably need to change the loop over `Ndof` to a triple loop over `Nqk`, `Nstate`, and `Nvertelem`. The last one probably isn't needed at compile time, though the first two we might want at compile time for unrolling.